saleor/payment/gateways/adyen/tests/webhooks/test_handle_notifications.py

```diff
@@ -989,3 +993,201 @@ def test_handle_cancel_or_refund_action_cancel_invalid_payment_id(
     handle_cancel_or_refund(notification, config)
 
     assert f"Unable to decode the payment ID {invalid_reference}." in caplog.text
+
+
+def test_handle_not_created_order_order_created(
+    checkout_ready_to_complete, payment_adyen_for_checkout, adyen_plugin, notification
+):
+    payment_adyen_for_checkout.charge_status = ChargeStatus.FULLY_CHARGED
+    payment_adyen_for_checkout.save(update_fields=["charge_status"])
+
+    adyen_plugin()
+    handle_not_created_order(
+        notification(),
+        payment_adyen_for_checkout,
+        payment_adyen_for_checkout.checkout,
+        TransactionKind.CAPTURE,
+        get_plugins_manager(),
+    )
+
+    payment_adyen_for_checkout.refresh_from_db()
+
+    assert payment_adyen_for_checkout.order
+
+
+@patch("saleor.payment.gateway.refund")
+@patch("saleor.checkout.complete_checkout._get_order_data")
+def test_handle_not_created_order_refund_when_create_order_raises(
+    order_data_mock, refund_mock, payment_adyen_for_checkout, adyen_plugin, notification
+):
+    order_data_mock.side_effect = ValidationError("Test error")
+
+    payment_adyen_for_checkout.charge_status = ChargeStatus.FULLY_CHARGED
+    payment_adyen_for_checkout.save(update_fields=["charge_status"])
+
+    adyen_plugin()
+    handle_not_created_order(
+        notification(),
+        payment_adyen_for_checkout,
+        payment_adyen_for_checkout.checkout,
+        TransactionKind.CAPTURE,
+        get_plugins_manager(),
+    )
+
+    assert payment_adyen_for_checkout.can_refund()
+    assert refund_mock.call_count == 2
```

Comment:
@L3str4nge why is 2 the expected count here?

Reply:
It's tricky because if you take a look on the implementation we actually raises ValidationError twice. It should be refactored,  we have notes about that.